Jetpack Plans: Add product API key to purchase page

ryelle opened the pull request:
Fixes #205, #10494

This PR starts the work to show product API keys in Calypso. It creates a new component card which is displayed on the Manage Purchase page (`/me/purchases`). This only shows up on paid Jetpack sites.

**To test**

1. On a Jetpack site with a Personal, Professional, or Premium plan, view your plan `/plans/my-plan/$site`
2. Click Manage Payment (this is the easiest way to get to the purchase screen)
3. Below your site info, there should be a second card with your API keys.

![screen shot 2017-03-16 at 4 31 31 pm](https://cloud.githubusercontent.com/assets/541093/24017362/16736e22-0a66-11e7-95da-a5ab8d17eb8f.png)
(those are not real keys for the screenshot, but it does show your actual keys 😄 )

cc @johnHackworth

ryelle commented:
This also triggers the "Your plan needs setting up" notice due to how it detects that from the state.
 
![screen shot 2017-03-16 at 4 36 29 pm](https://cloud.githubusercontent.com/assets/541093/24017539/d582a792-0a66-11e7-8a1d-38ea6a5acd66.png)

Maybe there should be a separate object in state for "started the install process"? Or we could create another query to fetch these instructions, that doesn't also set the state for auto-installing. Pinging @ebinnion since you've been working on this lately :)

johnHackworth commented:
this is amazing, @ryelle !!!

ebinnion commented:
Could we modify the selector to check if we never started installing? Perhaps something like checking if all plugins are `wait` status?

ryelle commented:
>  Perhaps something like checking if all plugins are wait status?

Ah, that's a good idea. I'll do that in another PR :)

ryelle commented:
I've addressed the blue banner issue in #12377

ryelle commented:
I've updated this PR to basically be groundwork for the new design (see #12746) so that we don't have monster PRs :) This one ended up a bit large because I also switched `ManagePurchase` to use localize instead of `this.translate`.

tyxla commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -681,6 +685,13 @@ const ManagePurchase = React.createClass( {
+				<PurchasePlanDetails
+					hasLoadedSites={ this.props.hasLoadedSites }
+					hasLoadedUserPurchasesFromServer={ this.props.hasLoadedUserPurchasesFromServer }
+					selectedSite={ this.props.selectedSite }
```
Does it make sense to omit these props here and add them the `connect()` of `PurchasePlanDetails` directly, since it's already connected anyway? This could help make the component more independent. Or are we doing it this way for consistency with other components like `RemovePurchase` for example?

tyxla commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -436,7 +440,7 @@ const ManagePurchase = React.createClass( {
 				showDismiss={ false }
 				status="is-error"
-				text={ this.translate( 'This purchase has expired and is no longer in use.' ) }>
+				text={ this.props.translate( 'This purchase has expired and is no longer in use.' ) }>
```
If we assign `const { translate } = this.props` here and there, we could save some code by avoiding to have to use `this.props.` everywhere.

tyxla commented on client/me/purchases/manage-purchase/plan-details/index.jsx:
```diff
@@ -0,0 +1,61 @@
+				<QueryPluginKeys siteId={ selectedSite.ID } />
+				{ pluginList.map( ( plugin, i ) => {
+					return (
+						<FormFieldset key={ i }>
```
A design item really, but there's some extra space below the last `FormFieldset`:

![](https://cldup.com/TBdDcsY3rQ.png)

Can we avoid that by setting a `margin-bottom: 0` of the `.form-fieldset:last-child`?

tyxla commented on client/me/purchases/manage-purchase/plan-details/index.jsx:
```diff
@@ -0,0 +1,61 @@
+import { isJetpackPlan, isFreeJetpackPlan } from 'lib/products-values';
+import { getPluginsForSite } from 'state/plugins/premium/selectors';
+
+class PurchasePlanDetails extends Component {
```
This component is to be used strictly for Jetpack plans, does it make sense to rename it to `PurchaseJetpackPlanDetails` or similar?

tyxla commented on client/me/purchases/manage-purchase/plan-details/index.jsx:
```diff
@@ -0,0 +1,61 @@
+	}
+
+	render() {
+		if ( isDataLoading( this.props ) || ! this.props.selectedSite ) {
```
Would it make sense to display a placeholder instead of nothing, for example in a situation where purchases and site have loaded, but plugin keys haven't?

tyxla commented on client/me/purchases/manage-purchase/plan-details/index.jsx:
```diff
@@ -0,0 +1,61 @@
+			return null;
+		}
+
+		const { selectedSite, pluginList } = this.props;
```
Let's move this line to the top of the method, this way we won't have to use `this.props.selectedSite` and will use just `selectedSite` instead.

tyxla commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -681,6 +685,13 @@ const ManagePurchase = React.createClass( {
 					{ contactSupportToRenewMessage }
 				</Card>
 
+				<PurchasePlanDetails
```
Can we completely avoid rendering this one if we know it's not a Jetpack site? `isJetpackSite` could come in handy.

tyxla commented on client/me/purchases/manage-purchase/plan-details/index.jsx:
```diff
@@ -0,0 +1,61 @@
+import FormFieldset from 'components/forms/form-fieldset';
+import FormLabel from 'components/forms/form-label';
+import QueryPluginKeys from 'components/data/query-plugin-keys';
+import { getPurchase, isDataLoading } from '../../utils';
```
I'd avoid relative paths at all costs, especially such with 2 directories above - they often break as we move files here and there. And `client/me/purchases/utils` isn't that long anyway 😉

ryelle commented:
Thanks @tyxla, I've addressed some of what you've said. The rest isn't really applicable because this is just a first-pass of adding these fields, the design will be updated in a following PR, along with some other fields that are used on a wp.com site (see #12746)

rickybanister commented:
It's a little hard to do a 'design review' of this since there's a bit of an order of operations problem with merging the other PRs to get the new design in place. 

Can we just put a feature flag on this card until the other PRs can be merged?

ryelle commented:
@rickybanister I've added a feature flag so this should only show up in development & wpcalypso. Are you OK with me merging this as-is?

rickybanister commented:
Yup, in that case let's merge so we can get everything starting to come together.
